### ISSUE TITLE
Broken with SQL Mode "ONLY_FULL_GROUP_BY"

### ISSUE THREAD

**Issue by bennyborn** [isotope/core#1886]
While deploying an Isotope installation (2.4.4, Contao 4.4.7) on a hosting running `MySQL 5.7.20` we encountered breaking problems with Isotope and SQL mode `ONLY_FULL_GROUP_BY`.

The following query origins from a product listing module and fails because since `MySQL 5.7.5` they enabled SQL mode `ONLY_FULL_GROUP_BY` by default and the field `c.sorting` is missing in the GROUP BY.

```sql
SELECT
	tl_iso_product.*, 
	'de' AS language, 
	tl_iso_product.images AS images_fallback, 
	c.sorting 
FROM tl_iso_product 
	LEFT OUTER JOIN tl_iso_product_category c ON tl_iso_product.id = c.pid 
WHERE
		tl_iso_product.language = ''
	AND tl_iso_product.published = '1' 
	AND ( tl_iso_product.start = '' OR tl_iso_product.start < '1510237440' ) 
	AND ( tl_iso_product.stop = '' OR tl_iso_product.stop > '1510237500' ) 
	AND tl_iso_product.id IN ( 14, 18, 19, 30 ) 
GROUP BY tl_iso_product.id 
ORDER BY Find_in_set(tl_iso_product.id, '14,18,19,30');
```
```
SQLSTATE[42000]: Syntax error or access violation: 1055 Expression #57 of SELECT list is not in GROUP BY clause and contains nonaggregated column 'notimportantdbname.c.sorting' which is not functionally dependent on columns in GROUP BY clause; this is incompatible with sql_mode=only_full_group_by
```

It seems that similar problems caused by this SQL mode are happening in almost all other Isotope frontend modules as well.

**Comment by electronicamystica** [comment]
I just ran into the same issue (same contao/isotope/mysql that you're using), when calling the iso_products backend module after adding the first product:

```sql
SELECT 
		tl_iso_product.*,
		'en' AS language, 
		IF(tl_iso_product.pid>0,
		parent.type, 
		tl_iso_product.type) AS type, 
		tl_iso_product.images AS images_fallback, 
		c.sorting 
FROM tl_iso_product 
		LEFT OUTER JOIN tl_iso_product_category c ON tl_iso_product.id=c.pid 
		LEFT OUTER JOIN tl_iso_product parent ON tl_iso_product.pid=parent.id 
WHERE 
			tl_iso_product.language='' 
		AND tl_iso_product.id='3'
		GROUP BY tl_iso_product.id LIMIT 0,1 
```

`SQLSTATE[42000]: Syntax error or access violation: 1055 Expression #35 of SELECT list is not in GROUP BY clause and contains nonaggregated column 'databasename.c.sorting' which is not functionally dependent on columns in GROUP BY clause; this is incompatible with sql_mode=only_full_group_by`

Any ideas for a workaround?

**Comment by fritzmg** [comment]
Related:

* https://github.com/contao/newsletter-bundle/issues/9
* https://github.com/terminal42/contao-leads/issues/49

**Comment by bennyborn** [comment]
> Any ideas for a workaround?

Since I have at least always one customer specific extension for each project I simply added a Hook where I overwrite the SQL mode for the session.

inside `config.php`
```php
$GLOBALS['TL_HOOKS']['initializeSystem'][] = array('numero2\Project\CustomClass', 'fixMySQLFullGroupBy');
```

inside `CustomClass.php`
```php
public function fixMySQLFullGroupBy() {

        $mode = NULL;
        $mode = \Database::getInstance()->execute("SELECT @@sql_mode AS mode;");

        if( !empty($mode->mode) && strpos($mode->mode,'ONLY_FULL_GROUP_BY') !== FALSE ) {

            $newMode = str_replace('ONLY_FULL_GROUP_BY','',$mode->mode);
            \Database::getInstance()->prepare("SET SESSION sql_mode = ?")->execute($newMode);
        }
    }
```

It's a hell of a hack but this way I can ensure I keep all the other enabled modes untouched while dropping the `ONLY_FULL_GROUP_BY`. Works like a charme though :D

**Comment by electronicamystica** [comment]
awesome!
I just removed the  `ONLY_FULL_GROUP_BY` from my /etc/mysql/my.cnf which also did the trick. I like your solution much more though, as it doesn't affect the other installations and works without root:-)
cheers!

**Comment by felixpfeiffer** [comment]
Any news if this will be fixed in Isotope or is using the hook the only way to fix it for next time?

**Comment by aschempp** [comment]
I have applied multiple patches to fix this in the `hotfix/2.5.5` branch. Is anyone able to test that and provide feedback or cases that (still) do not work?

**Comment by aschempp** [comment]
I've release 2.5.5 to fix this
